File: app/src/main/java/com/cookplan/shopping_list/total_list/TotalShoppingListPresenterImpl.java
package com.cookplan.shopping_list.total_list;


import com.cookplan.RApplication;
import com.cookplan.models.CookPlanError;
import com.cookplan.models.Ingredient;
import com.cookplan.models.MeasureUnit;
import com.cookplan.models.ProductCategory;
import com.cookplan.models.ShopListStatus;
import com.cookplan.shopping_list.ShoppingListBasePresenterImpl;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.CompletableObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by DariaEfimova on 24.03.17.
 */

public class TotalShoppingListPresenterImpl extends ShoppingListBasePresenterImpl implements TotalShoppingListPresenter {

    private TotalShoppingListView mainView;
    private Map<String, List<Ingredient>> ProductToIngredientMap;

    public TotalShoppingListPresenterImpl(TotalShoppingListView mainView) {
        super();
        this.mainView = mainView;
    }

    @Override
    protected void setError(String message) {
        if (mainView != null) {
            mainView.setErrorToast(message);
        }
    }

    @Override
    public void sortIngredientList(List<Ingredient> userIngredients) {

        //fill the map
        ProductToIngredientMap = new HashMap<>();
        for (Ingredient ingredient : userIngredients) {
            if (ingredient.getShopListStatus() != ShopListStatus.NONE) {
                ArrayList<Ingredient> ingredients;
                if (ProductToIngredientMap.containsKey(ingredient.getName())
                        && ProductToIngredientMap.get(ingredient.getName()) != null) {
                    ingredients = (ArrayList<Ingredient>) ProductToIngredientMap.get(ingredient.getName());
                } else {
                    ingredients = new ArrayList<>();
                }
                ingredients.add(ingredient);
                ProductToIngredientMap.put(ingredient.getName(), ingredients);
            }
        }

        //calculate amounts
        List<Ingredient> needToBuyIngredients = new ArrayList<>();
        for (Map.Entry<String, List<Ingredient>> entry : ProductToIngredientMap.entrySet()) {
            Ingredient needToBuyIng = calculateAmountForEntry(entry, ShopListStatus.NEED_TO_BUY);
            if (needToBuyIng != null) {
                needToBuyIngredients.add(needToBuyIng);
            }

            Ingredient alreadyBoughtIngred = calculateAmountForEntry(entry, ShopListStatus.ALREADY_BOUGHT);
            if (alreadyBoughtIngred != null) {
                needToBuyIngredients.add(alreadyBoughtIngred);
            }
        }
        List<Ingredient> needToBuySortedIngredients = new ArrayList<>();
        for (ProductCategory category : RApplication.getPriorityList()) {
            for (Ingredient ingredient : needToBuyIngredients) {
                if (ingredient.getCategory() == category) {
                    needToBuySortedIngredients.add(ingredient);
                }
            }
        }
        if (mainView != null) {
            if (ProductToIngredientMap.size() != 0) {
                mainView.setIngredientLists(needToBuySortedIngredients);
            } else {
                mainView.setEmptyView();
            }
        }
    }

    private Ingredient calculateAmountForEntry(Map.Entry<String, List<Ingredient>> entry, ShopListStatus status) {
        String productName = entry.getKey();
        List<Ingredient> ingredients = entry.getValue();
        if (!ingredients.isEmpty()) {
            ProductCategory category = ingredients.get(0).getCategory();
            Map<MeasureUnit, Double> shopMap = new HashMap<>();
            Map<MeasureUnit, Double> restMap = new HashMap<>();
            for (Ingredient ingredient : ingredients) {
                if (ingredient.getShopListStatus() == status
                        && ingredient.getShopMeasureList() != null
                        && ingredient.getShopAmountList() != null) {
                    double localAmount;
                    MeasureUnit unit = null;
                    double amount = -1.;
                    for (int i = 0; i < ingredient.getShopMeasureList().size(); i++) {
                        MeasureUnit shopUnit = ingredient.getShopMeasureList().get(i);
                        double shopAmount = ingredient.getShopAmountList().get(i);
                        if (shopAmount > 1e-8) {
                            unit = shopUnit;
                            amount = shopAmount;
                            if (shopMap.containsKey(unit)) {
                                localAmount = shopMap.get(unit) + amount;
                            } else {
                                localAmount = amount;
                            }
                            shopMap.put(unit, localAmount);
                        }
                    }

                    if (amount < 1e-8 && unit == null) {
                        unit = ingredient.getMainMeasureUnit();
                        amount = ingredient.getMainAmount();
                        if (restMap.containsKey(unit)) {
                            localAmount = restMap.get(unit) + amount;
                        } else {
                            localAmount = amount;
                        }
                        restMap.put(unit, localAmount);
                    }
                }
            }

            if (restMap.isEmpty() && shopMap.isEmpty()) {
                return null;
            }
            String shopAmountString = "";
            for (Map.Entry<MeasureUnit, Double> shopEntry : shopMap.entrySet()) {
                String string = shopEntry.getKey().toValueString(shopEntry.getValue());
                if (shopAmountString.isEmpty()) {
                    shopAmountString = string;
                } else {
                    shopAmountString = shopAmountString + " (" + string + ")";
                }
            }
            String restAmountString = "";
            for (Map.Entry<MeasureUnit, Double> measureEntry : restMap.entrySet()) {
                String string = measureEntry.getKey().toValueString(measureEntry.getValue());
                restAmountString = restAmountString.isEmpty() ? string : restAmountString + " + " + string;
            }
            if (!restAmountString.isEmpty()) {
                shopAmountString = shopAmountString + " + " + restAmountString;
            }
            return new Ingredient(productName, shopAmountString, status, category);
        } else {
            return null;
        }
    }


    @Override
    public void changeShopListStatus(Ingredient totalIngredient, ShopListStatus newStatus) {
        List<Ingredient> ingredientList = ProductToIngredientMap.get(totalIngredient.getName());
        for (Ingredient realIngredient : ingredientList) {
            realIngredient.setShopListStatus(newStatus);
            ingredientDataProvider.updateShopStatus(realIngredient)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new CompletableObserver() {
                        @Override
                        public void onSubscribe(Disposable d) {

                        }

                        @Override
                        public void onComplete() {

                        }

                        @Override
                        public void onError(Throwable e) {
                            if (mainView != null) {
                                mainView.setErrorToast(e.getMessage());
                            }
                        }
                    });
        }
    }

    @Override
    public void deleteIngredients(List<Ingredient> localIngredients) {
        for (Ingredient totalIngredient : localIngredients) {
            ShopListStatus status = totalIngredient.getShopListStatus();
            List<Ingredient> realIngredients = ProductToIngredientMap.get(totalIngredient.getName());
            for (Ingredient realIngredient : realIngredients) {
                if (status == realIngredient.getShopListStatus()) {
                    realIngredient.setShopListStatus(ShopListStatus.NONE);
                    if (realIngredient.getRecipeId() == null) {
                        ingredientDataProvider.removeIngredient(realIngredient)
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe(new CompletableObserver() {
                                    @Override
                                    public void onSubscribe(Disposable d) {

                                    }

                                    @Override
                                    public void onComplete() {

                                    }

                                    @Override
                                    public void onError(Throwable e) {
                                        if (mainView != null && e instanceof CookPlanError) {
                                            mainView.setErrorToast(e.getMessage());
                                        }
                                    }
                                });
                    } else {
                        ingredientDataProvider.updateShopStatus(realIngredient)
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe(new CompletableObserver() {
                                    @Override
                                    public void onSubscribe(Disposable d) {

                                    }

                                    @Override
                                    public void onComplete() {

                                    }

                                    @Override
                                    public void onError(Throwable e) {
                                        if (mainView != null) {
                                            mainView.setErrorToast(e.getMessage());
                                        }
                                    }
                                });
                    }
                }
            }
        }
    }
}

File: app/src/main/java/com/cookplan/shopping_list/total_list/TotalShoppingListPresenterImpl.kt
package com.cookplan.shopping_list.total_list


import com.cookplan.RApplication
import com.cookplan.models.Ingredient
import com.cookplan.models.MeasureUnit
import com.cookplan.models.ShopListStatus
import com.cookplan.shopping_list.ShoppingListBasePresenterImpl
import io.reactivex.CompletableObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlin.collections.Map.Entry

/**
 * Created by DariaEfimova on 24.03.17.
 */

class TotalShoppingListPresenterImpl(private val mainView: TotalShoppingListView?) : ShoppingListBasePresenterImpl(), TotalShoppingListPresenter {
    private var ProductToIngredientMap: HashMap<String, List<Ingredient>> = hashMapOf()

    override fun setError(message: String?) {
        mainView?.setErrorToast(message ?: "")
    }

    override fun sortIngredientList(userIngredients: List<Ingredient>) {

        //fill the map
        ProductToIngredientMap.clear()
        for (ingredient in userIngredients) {
            if (ingredient.shopListStatus !== ShopListStatus.NONE) {
                val ingredients = (ProductToIngredientMap.get(ingredient.name) ?: arrayListOf()) as ArrayList<Ingredient>

                ingredients.add(ingredient)
                ProductToIngredientMap.put(ingredient.name ?: String(), ingredients)
            }
        }

        //calculate amounts
        val needToBuyIngredients = ArrayList<Ingredient>()
        for (entry in ProductToIngredientMap.entries) {
            val needToBuyIng = getShopListIngredient(entry, ShopListStatus.NEED_TO_BUY)
            if (needToBuyIng != null) {
                needToBuyIngredients.add(needToBuyIng)
            }

            val alreadyBoughtIngred = getShopListIngredient(entry, ShopListStatus.ALREADY_BOUGHT)
            if (alreadyBoughtIngred != null) {
                needToBuyIngredients.add(alreadyBoughtIngred)
            }
        }
        val needToBuySortedIngredients = ArrayList<Ingredient>()
        for (category in RApplication.getPriorityList()) {
            for (ingredient in needToBuyIngredients) {
                if (ingredient.category === category) {
                    needToBuySortedIngredients.add(ingredient)
                }
            }
        }
        if (ProductToIngredientMap.size != 0) {
            mainView?.setIngredientLists(needToBuySortedIngredients)
        } else {
            mainView?.setEmptyView()
        }
    }

    private fun getShopListIngredient(entry: Entry<String, List<Ingredient>>, status: ShopListStatus): Ingredient? {
        val productName = entry.key
        val ingredients = entry.value
        if (!ingredients.isEmpty()) {
            val shopMap = HashMap<MeasureUnit, Double>()
            val restMap = HashMap<MeasureUnit, Double>()
            for (ingredient in ingredients) {
                if (ingredient.shopListStatus === status) {
                    for (i in 0 until ingredient.shopMeasureList.size) {
                        val shopUnit = ingredient.shopMeasureList.get(i)
                        val shopAmount = ingredient.shopAmountList.get(i)
                        if (shopAmount > 1e-8) {
                            var mapAmount = shopMap.get(shopUnit) ?: 0.toDouble()
                            var localAmount = mapAmount + shopAmount
                            shopMap.put(shopUnit, localAmount)
                        }
                    }

                    if (shopMap.isEmpty()) {
                        val restUnit = ingredient.mainMeasureUnit
                        val restAmount = ingredient.mainAmount
                        var mapAmount = restMap.get(restUnit) ?: 0.toDouble()
                        var localAmount = mapAmount + restAmount
                        restMap.put(restUnit, localAmount)
                    }
                }
            }

            if (restMap.isEmpty() && shopMap.isEmpty()) {
                return null
            }
            var shopAmountString = ""
            for ((key, value) in shopMap) {
                val string = key.toStringForShopList(value)
                if (shopAmountString.isEmpty()) {
                    shopAmountString = string
                } else {
                    shopAmountString = "$shopAmountString ($string)"
                }
            }
            var restAmountString = ""
            for ((key, value) in restMap) {
                val string = key.toStringForShopList(value)
                restAmountString = if (restAmountString.isEmpty()) string else restAmountString + " + " + string
            }
            if (!restAmountString.isEmpty()) {
                shopAmountString = shopAmountString + " + " + restAmountString
            }
            val category = ingredients.get(0).category
            return Ingredient(productName, shopAmountString, status, category)
        } else {
            return null
        }
    }


    override fun changeShopListStatus(ingredient: Ingredient, newStatus: ShopListStatus) {
        val ingredientList = ProductToIngredientMap.get(ingredient.name) ?: listOf()
        for (realIngredient in ingredientList) {
            realIngredient.shopListStatus = newStatus
            ingredientDataProvider.updateShopStatus(realIngredient)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(object : CompletableObserver {
                        override fun onSubscribe(d: Disposable) {

                        }

                        override fun onComplete() {

                        }

                        override fun onError(e: Throwable) {
                            mainView?.setErrorToast(e.message ?: "")
                        }
                    })
        }
    }

    override fun deleteIngredients(ingredients: List<Ingredient>) {
        for ((_, name, _, _, _, _, _, _, _, status) in ingredients) {
            val realIngredients = ProductToIngredientMap.get(name) ?: listOf()
            for (realIngredient in realIngredients) {
                if (status === realIngredient.shopListStatus) {
                    realIngredient.shopListStatus = ShopListStatus.NONE
                    if (realIngredient.recipeId == null) {
                        ingredientDataProvider.removeIngredient(realIngredient)
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe(object : CompletableObserver {
                                    override fun onSubscribe(d: Disposable) {

                                    }

                                    override fun onComplete() {

                                    }

                                    override fun onError(e: Throwable) {
                                        mainView?.setErrorToast(e.message ?: "")
                                    }
                                })
                    } else {
                        ingredientDataProvider.updateShopStatus(realIngredient)
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe(object : CompletableObserver {
                                    override fun onSubscribe(d: Disposable) {

                                    }

                                    override fun onComplete() {

                                    }

                                    override fun onError(e: Throwable) {
                                        mainView?.setErrorToast(e.message ?: "")
                                    }
                                })
                    }
                }
            }
        }
    }
}